services/index/src/helpers.js

```diff
@@ -1,111 +1,360 @@
+    let ns = task.namespace + '.' + task.name;
+    // Remove separate if there is no need
+    if (task.namespace.length === 0 || task.name.length === 0) {
+      ns = task.namespace + task.name;
     }
 
-    // Create namespace hierarchy
-    return options.Namespace.ensureNamespace(
-      namespace,
-      expires,
-    ).then(function() {
-      return options.IndexedTask.create({
-        namespace: namespace,
-        name: name,
-        rank: input.rank,
-        taskId: input.taskId,
-        data: input.data,
-        expires: expires,
+    return {
+      namespace: ns,
+      taskId: task.taskId,
+      rank: task.rank,
+      data: _.cloneDeep(task.data),
+      expires: task.expires.toJSON(),
+    };
+  },
+  /**
+   * Insert task into `namespace` where:
+   *
+   * input:
+   * {
+   *   expires:      // Date object
+   *   data:         // IndexedTask.data or JSON date string
+   *   taskId:       // TaskId for task
+   *   rank:         // IndexedTask.rank
+   * }
+   *
+   */
+  insertTask(db, namespace, input) {
+    // Validate input
+    assert(input.expires instanceof Date, 'expires must be a Date object');
+    assert(input.data instanceof Object, 'data must be an object');
+    assert(input.taskId, 'taskId must be given');
+    assert(typeof input.rank === 'number', 'rank must be a number');
+    assert(db,
+      'db must be set');
+
+    // Get namespace and ensure that we have a least one dot
+    namespace = namespace.split('.');
+
+    // Find name and namespace
+    let name = namespace.pop() || '';
+    namespace = namespace.join('.');
+
+    // Find expiration time and parse as date object
+    let expires = new Date(input.expires);
+
+    // Attempt to load indexed task
+    return db.fns.get_indexed_task(namespace, name)
+      .then(async function(t) {
+        const task = exports.taskUtils.fromDbRows(t);
+
+        // Update if we prefer input over what we have
+        if (task.rank <= input.rank) {
+          const updatedTask = await db.fns.update_indexed_task(
+            task.namespace,
+            task.name,
+            input.rank,
+            input.taskId,
+            input.data,
+            expires,
+            task.etag,
+          );
+
+          await exports.namespaceUtils.ensureNamespace(db, namespace, expires);
+
+          return exports.taskUtils.fromDbRows(updatedTask);
+        } else {
+          return task;
+        }
+      }, function(err) {
+        // Re-throw error, if it's not a 404
+        if (err.code !== 'P0002') {
+          throw err;
+        }
+
+        // Create namespace hierarchy
+        return exports.namespaceUtils.ensureNamespace(
+          db,
+          namespace,
+          expires,
+        ).then(async function() {
+          const etag = (await db.fns.create_indexed_task(
+            namespace,
+            name,
+            input.rank,
+            input.taskId,
+            input.data,
+            expires,
+          ))[0].create_indexed_task;
```

Comment:
It looks like all of this is kind of an elaborate "upsert" operation, with complications that we only want to replace an existing entity if its rank is lower, and that we need to ensure the namespace exists if it does not already.  I suspect all of that could be done more simply directly in the DB, but maybe that's a bit of a heavy lift for now -- we could leave that to a follow-up.